caddyhttp/digestauth/digestauth_test.go

```diff
@@ -0,0 +1,244 @@
+			} else {
+				if req.Header.Get("Authorization") == "" {
+					// see issue #1508: https://github.com/mholt/caddy/issues/1508
+					t.Errorf("Test %d: Expected Authorization header to be retained after successful auth, but was empty", i)
+				}
+			}
+		}
+	}
+}
+
+func TestMultipleOverlappingRules(t *testing.T) {
+	rw := DigestAuth{
+		Next: httpserver.HandlerFunc(contentHandler),
+		Rules: []Rule{
+			{Users: NewSimpleUserStore(map[string]string{"t": "p1"}),
+				Resources: []string{"/t"}, Realm: "Restricted"},
+			{Users: NewSimpleUserStore(map[string]string{"t1": "p2"}),
+				Resources: []string{"/t/t"}, Realm: "Restricted"},
+		},
+	}
+	rw.Rules[0].Digester = NewDigestHandler(rw.Rules[0].Realm, nil, nil, rw.Rules[0].Users)
+	rw.Rules[1].Digester = NewDigestHandler(rw.Rules[1].Realm, nil, nil, rw.Rules[1].Users)
+
+	tests := []struct {
+		from   string
+		result int
+		cred   string
+	}{
+		{"/t", http.StatusOK, "t:p1"},
+		{"/t/t", http.StatusOK, "t:p1"},
+		{"/t/t", http.StatusOK, "t1:p2"},
+		{"/a", http.StatusOK, "t1:p2"},
+		{"/t/t", http.StatusUnauthorized, "t1:p3"},
+		{"/t", http.StatusUnauthorized, "t1:p2"},
+	}
+
+	for i, test := range tests {
+		req, err := http.NewRequest("GET", test.from, nil)
+		if err != nil {
+			t.Fatalf("Test %d: Could not create HTTP request %v", i, err)
+		}
+		auth := "Digest " + base64.StdEncoding.EncodeToString([]byte(test.cred))
+		req.Header.Set("Authorization", auth)
+
+		rec := httptest.NewRecorder()
+		result, err := rw.ServeHTTP(rec, req)
+		if err != nil {
+			t.Fatalf("Test %d: Could not ServeHTTP %v", i, err)
+		}
+		if result != test.result {
+			t.Errorf("Test %d: Expected Header '%d' but was '%d'",
+				i, test.result, result)
+		}
+	}
+
+}
+
+func contentHandler(w http.ResponseWriter, r *http.Request) (int, error) {
+	fmt.Fprintf(w, r.URL.String())
+	return http.StatusOK, nil
+}
+
+func TestHtpasswd(t *testing.T) {
+	//htpasswdPasswd := "IedFOuGmTpT8"
+	//htpasswdFile := `sha1:{SHA}dcAUljwz99qFjYR0YLTXx0RqLww=
+	//md5:$apr1$l42y8rex$pOA2VJ0x/0TwaFeAF9nX61`
+
+	//htfh, err := ioutil.TempFile("", "basicauth-")
+	//if err != nil {
+	//t.Skipf("Error creating temp file (%v), will skip htpassword test")
+	//return
+	//}
+	//defer os.Remove(htfh.Name())
+	//if _, err = htfh.Write([]byte(htpasswdFile)); err != nil {
+	//t.Fatalf("write htpasswd file %q: %v", htfh.Name(), err)
+	//}
+	//htfh.Close()
+
+	//for i, username := range []string{"sha1", "md5"} {
+	//rule := Rule{Username: username, Resources: []string{"/testing"}}
+
+	//siteRoot := filepath.Dir(htfh.Name())
+	//filename := filepath.Base(htfh.Name())
+	//if rule.Password, err = GetHtpasswdMatcher(filename, rule.Username, siteRoot); err != nil {
+	//t.Fatalf("GetHtpasswdMatcher(%q, %q): %v", htfh.Name(), rule.Username, err)
+	//}
+	//t.Logf("%d. username=%q", i, rule.Username)
+	//if !rule.Password(htpasswdPasswd) || rule.Password(htpasswdPasswd+"!") {
+	//t.Errorf("%d (%s) password does not match.", i, rule.Username)
+	//}
+	//}
+}
```

Comment:
Is this related to the other commented tests?

Reply:
^ same